File: frontend/src/types/alert.ts
/**
 * Alert types based on backend schema
 */

export type AlertCategory = 'irrigation' | 'soil' | 'pests' | 'health' | 'production' | 'system';
export type AlertSeverity = 'critical' | 'warning' | 'info';
export type AlertSource = 'sensor' | 'vision' | 'manual' | 'system';

export interface Alert {
  id: string;
  organization_id: string;
  farm_id: string | null;
  plot_id: string | null;
  row_id: string | null;
  tree_id: string | null;
  category: AlertCategory;
  severity: AlertSeverity;
  type: string;
  title: string;
  message: string;
  impact: string | null;
  suggested_action: string | null;
  source: AlertSource | null;
  source_id: string | null;
  timestamp: string;
  acknowledged_at: string | null;
  acknowledged_by: string | null;
  resolved_at: string | null;
  resolved_by: string | null;
  resolution_notes: string | null;
  recurrence_count: number;
  extra_data: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  // Joined fields
  plot_name?: string;
  farm_name?: string;
}

export interface AlertCreate {
  farm_id?: string;
  plot_id?: string;
  row_id?: string;
  tree_id?: string;
  category: AlertCategory;
  severity: AlertSeverity;
  type: string;
  title: string;
  message: string;
  impact?: string;
  suggested_action?: string;
  source?: AlertSource;
  source_id?: string;
  extra_data?: Record<string, unknown>;
}

export interface AlertAcknowledge {
  notes?: string;
}

export interface AlertResolve {
  resolution_notes?: string;
}

export interface AlertFilters {
  farm_id?: string;
  plot_id?: string;
  category?: AlertCategory;
  severity?: AlertSeverity;
  resolved?: boolean;
  acknowledged?: boolean;
}

File: backend/app/schemas/alert.py
"""Schemas de alertas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class AlertBase(BaseModel):
    """Schema base de alerta."""

    farm_id: UUID | None = None
    plot_id: UUID | None = None
    row_id: UUID | None = None
    tree_id: UUID | None = None
    category: str
    severity: str
    type: str
    title: str
    message: str
    impact: str | None = None
    suggested_action: str | None = None
    source: str | None = None
    source_id: UUID | None = None
    extra_data: dict = {}


class AlertCreate(AlertBase):
    """Schema para criacao de alerta."""

    pass


class AlertUpdate(BaseModel):
    """Schema para atualizacao de alerta."""

    category: str | None = None
    severity: str | None = None
    type: str | None = None
    title: str | None = None
    message: str | None = None
    impact: str | None = None
    suggested_action: str | None = None
    extra_data: dict | None = None


class AlertAcknowledge(BaseModel):
    """Schema para reconhecimento de alerta."""

    notes: str | None = None


class AlertResolve(BaseModel):
    """Schema para resolucao de alerta."""

    resolution_notes: str | None = None


class AlertResponse(AlertBase):
    """Schema de resposta de alerta."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    timestamp: datetime
    acknowledged_at: datetime | None = None
    acknowledged_by: UUID | None = None
    resolved_at: datetime | None = None
    resolved_by: UUID | None = None
    resolution_notes: str | None = None
    recurrence_count: int = 0
    created_at: datetime
    updated_at: datetime
